lib/cylc/network/pyro_base.py

```diff
@@ -124,9 +124,36 @@ def _set_uri(self):
             "suite": self.suite,
             "owner": self.owner,
             "target": self.target_server_object}
-        port_file_path = os.path.join(
-            GLOBAL_CFG.get(['pyro', 'ports directory']), self.suite)
+
+        if ((self.host is None or self.port is None) and
+                'CYLC_SUITE_RUN_DIR' in os.environ):
+            # Looks like we are in a running task job, so we should be able to
+            # use "cylc-suite-env" file under the suite running directory
+            env_file = os.path.join(
+                os.environ['CYLC_SUITE_RUN_DIR'], 'cylc-suite-env')
+            try:
+                suite, host, port = None, None, None
+                for line in open(env_file):
+                    key, value = line.strip().split('=', 1)
+                    if key == 'CYLC_SUITE_NAME' and value == self.suite:
+                        # This ensures we have the correct suite
+                        suite = value
+                    elif key == 'CYLC_SUITE_HOST':
+                        host = value
+                    elif key == 'CYLC_SUITE_PORT':
+                        port = value
+                if suite and host and port:
+                    # This ensures that all values are good
+                    uri_data['host'] = host
+                    uri_data['port'] = port
+                    self.host = host
+                    self.port = port
+            except (IOError, ValueError):
+                pass
```

Comment:
The `cylc-suite-env` file is also read and parsed in `lib/cylc/task_message`, but with different code.  Maybe time for a small module that handles parsing (and writing) the file, just returning a dict of parsed values?